### ISSUE TITLE
Coordinate conversion between WGS84 (lon-lat) and Web Mercator projection (meters)

### ISSUE THREAD

**Issue by navispatial** [navibyte/geospatial#55]
See `transform` and `project` functions introduced in #15. See some info about coordinate reference systems in #51. See also #54 and #49.

**Comment by navispatial** [comment]
Definition refined:

```dart
/// Returns a function that projects geographic points to web mercator points.
/// 
/// By default, target points of [R] are created using [factory]. 
///
/// The returned conversion function expects source coordinates to be in the
/// `CRS84` (that is 'WGS 84 longitude-latitude' or EPSG:4326) coordinate
/// reference system. Source `lon` and `lat` coordinates are converted to target
/// `x` and `y` coordinates of the Web Mercator (EPSG:3857) projection.
///
/// Other coordinates, if available in the source and if expected to target
/// coordinates, are then copied (`elev` => `z` and `m` => `m`) without any
/// changes.
ProjectPoint<GeoPoint, R> wgs84ToWebMercator<R extends CartesianPoint>(
  PointFactory<R> factory,
);

/// Returns a function that unprojects web mercator points to geographic points.
/// 
/// By default, target points of [R] are created using [factory].
///
/// The returned conversion function expects source coordinates to be in the
/// Web Mercator (EPSG:3857) projection. Source `x` and `y` coordinates are
/// converted to target `lon` and `lat` coordinates of the `CRS84` (that is
/// 'WGS 84 longitude-latitude' or EPSG:4326) coordinate reference system.
///
/// Other coordinates, if available in the source and if expected to target
/// coordinates, are then copied  (`z` => `elev` and `m` => `m`) without any
/// changes.
ProjectPoint<CartesianPoint, R> webMercatorToWgs84<R extends GeoPoint>(
  PointFactory<R> factory,
);
```

**Comment by navispatial** [comment]
Implementation (based on [The Google Maps / Bing Maps Spherical Mercator Projection](https://alastaira.wordpress.com/2011/01/23/the-google-maps-bing-maps-spherical-mercator-projection/) by Alastair Aitchison)

```dart
ProjectPoint<GeoPoint, R> wgs84ToWebMercator<R extends CartesianPoint>(
  PointFactory<R> factory,
) =>
    (GeoPoint source, [PointFactory<R>? f]) {
      // project (lon, lat) to (x, y)
      final x = source.lon * 20037508.34 / 180.0;
      final y0 = math.log(math.tan((90.0 + source.lat) * math.pi / 360.0)) /
          (math.pi / 180.0);
      final y = y0 * 20037508.34 / 180;
      // return a projected point with optional z and m coordinates unchanged
      return (f ?? factory).newWith(
        x: x,
        y: y,
        z: source.z,
        m: source.m,
      );
    };

ProjectPoint<CartesianPoint, R> webMercatorToWgs84<R extends GeoPoint>(
  PointFactory<R> factory,
) =>
    (CartesianPoint source, [PointFactory<R>? f]) {
      // unproject (x, y) to (lon, lat)
      final lon = (source.x / 20037508.34) * 180.0;
      final lat0 = (source.y / 20037508.34) * 180.0;
      final lat = 180.0 /
          math.pi *
          (2 * math.atan(math.exp(lat0 * math.pi / 180.0)) - math.pi / 2);
      // return an unprojected point with optional z and m coordinates unchanged
      return (f ?? factory).newWith(
        x: lon,
        y: lat,
        z: source.z,
        m: source.m,
      );
    };
```

**Comment by navispatial** [comment]
Implemented in pre-release 0.8.0-a.4 as follows:

```dart
ProjectPoint<R> wgs84ToWebMercator<R extends CartesianPoint>(
  PointFactory<R> factory,
) =>
    (Point source, {PointFactory<R>? to}) {
      // source coordinates
      final lon = source.x; // longitude is at x
      final lat = source.y; // latitude is at y

      // project (lon, lat) to (x, y)
      final x = lon * 20037508.34 / 180.0;
      final y0 = math.log(math.tan((90.0 + lat) * math.pi / 360.0)) /
          (math.pi / 180.0);
      final y = y0 * 20037508.34 / 180;

      // return a projected point with optional z and m coordinates unchanged
      return (to ?? factory).newWith(
        x: x,
        y: y,
        z: source.z,
        m: source.m,
      );
    };

ProjectPoint<R> webMercatorToWgs84<R extends GeoPoint>(
  PointFactory<R> factory,
) =>
    (Point source, {PointFactory<R>? to}) {
      // source coordinates
      final x = source.x;
      final y = source.y;

      // unproject (x, y) to (lon, lat)
      final lon = (x / 20037508.34) * 180.0;
      final lat0 = (y / 20037508.34) * 180.0;
      final lat = 180.0 /
          math.pi *
          (2 * math.atan(math.exp(lat0 * math.pi / 180.0)) - math.pi / 2);

      // return an unprojected point with optional z and m coordinates unchanged
      return (to ?? factory).newWith(
        x: lon,
        y: lat,
        z: source.z,
        m: source.m,
      );
    };
```

**Comment by navispatial** [comment]
The current implementation as of [geobase 0.5.1](https://pub.dev/packages/geobase/versions/0.5.1)

[WGS84](https://github.com/navibyte/geospatial/blob/v0.12.0/dart/geobase/lib/src/projections/wgs84/wgs84.dart):

```dart
import '/src/coordinates/projection/projection_adapter.dart';

import 'web_mercator_projection.dart';

/// Projections for the WGS 84 geographic coordinate system.
class WGS84 {
  /// A projection adapter between WGS84 geographic and Web Mercator positions.
  ///
  /// Use `forward` of the adapter to return a projection for:
  /// * source: `lon` and `lat` geographic coordinates (WGS 84)
  /// * target: `x` and `y` coordinates ("EPSG:3857", WGS 84 / Web Mercator)
  ///
  /// Use `inverse` of the adapter to return a projection for:
  /// * source: `x` and `y` coordinates ("EPSG:3857", WGS 84 / Web Mercator)
  /// * target: `lon` and `lat` geographic coordinates (WGS 84)
  ///
  /// Other coordinates, if available in the source and if expected for target
  /// coordinates, are just copied (`elev` <=> `z` and `m` <=> `m`) without any
  /// changes.
  static const ProjectionAdapter webMercator = Wgs84ToWebMercatorAdapter();
}
```

[Wgs84ToWebMercatorAdapter](https://github.com/navibyte/geospatial/blob/v0.12.0/dart/geobase/lib/src/projections/wgs84/web_mercator_projection.dart):

```dart
import 'package:meta/meta.dart';

import '/src/codes/coords.dart';
import '/src/coordinates/base/position.dart';
import '/src/coordinates/crs/coord_ref_sys.dart';
import '/src/coordinates/projection/projection.dart';
import '/src/coordinates/projection/projection_adapter.dart';
import '/src/utils/format_validation.dart';
import '/src/utils/web_mercator_converter.dart';

// More information about WGS 84 and Web Mercator
// https://epsg.io/3857
// https://en.wikipedia.org/wiki/Web_Mercator_projection
// https://alastaira.wordpress.com/2011/01/23/the-google-maps-bing-maps-spherical-mercator-projection/
// https://docs.microsoft.com/en-us/bingmaps/articles/bing-maps-tile-system
// https://athene-forschung.unibw.de/doc/132233/132233.pdf (Some Principles of Web Mercator Maps and their Computation)

// A local helper class for conversions.
const _converter = WebMercatorConverter.epsg3857();

/// A projection adapter between WGS84 geographic and Web Mercator positions.
@internal
class Wgs84ToWebMercatorAdapter with ProjectionAdapter {
  /// A projection adapter between WGS84 geographic and Web Mercator positions.
  const Wgs84ToWebMercatorAdapter();

  @override
  CoordRefSys get sourceCrs => CoordRefSys.CRS84;

  @override
  CoordRefSys get targetCrs => CoordRefSys.EPSG_3857;

  @override
  Projection get forward => const _Wgs84ToWebMercatorProjection();

  @override
  Projection get inverse => const _WebMercatorToWgs84Projection();
}

class _Wgs84ToWebMercatorProjection with Projection {
  const _Wgs84ToWebMercatorProjection();

  @override
  R project<R extends Position>(
    Position source, {
    required CreatePosition<R> to,
  }) {
    // source coordinates
    final lon = source.x; // longitude at x
    final lat = source.y; // latitude at y

    // return a projected position
    return to.call(
      // project (lon, lat) to (x, y)
      x: _converter.toProjectedX(lon),
      y: _converter.toProjectedY(lat),
      // optional z and m coordinates unchanged
      z: source.optZ,
      m: source.optM,
    );
  }

  @override
  List<double> projectCoords(
    Iterable<double> source, {
    List<double>? target,
    required Coords type,
  }) {
    final dim = type.coordinateDimension;
    final result = target ?? List<double>.filled(source.length, 0.0);

    var offset = 0;
    final iter = source.iterator;
    while (iter.moveNext()) {
      // project (lon, lat) to (x, y)
      result[offset] = _converter.toProjectedX(iter.current);
      result[offset + 1] = iter.moveNext()
          ? _converter.toProjectedY(iter.current)
          : throw invalidCoordinates;
      // optional z and m coordinates unchanged
      if (dim >= 3) {
        result[offset + 2] =
            iter.moveNext() ? iter.current : throw invalidCoordinates;
      }
      if (dim >= 4) {
        result[offset + 3] =
            iter.moveNext() ? iter.current : throw invalidCoordinates;
      }
      offset += dim;
    }

    return result;
  }
}

class _WebMercatorToWgs84Projection with Projection {
  const _WebMercatorToWgs84Projection();

  @override
  R project<R extends Position>(
    Position source, {
    required CreatePosition<R> to,
  }) {
    // return an unprojected position
    return to.call(
      // unproject (x, y) to (lon, lat)
      x: _converter.fromProjectedX(source.x),
      y: _converter.fromProjectedY(source.y),
      // optional z and m coords unchanged
      z: source.optZ,
      m: source.optM,
    );
  }

  @override
  List<double> projectCoords(
    Iterable<double> source, {
    List<double>? target,
    required Coords type,
  }) {
    final dim = type.coordinateDimension;
    final result = target ?? List<double>.filled(source.length, 0.0);

    var offset = 0;
    final iter = source.iterator;
    while (iter.moveNext()) {
      // unproject (x, y) to (lon, lat)
      result[offset] = _converter.fromProjectedX(iter.current);
      result[offset + 1] = iter.moveNext()
          ? _converter.fromProjectedY(iter.current)
          : throw invalidCoordinates;
      // optional z and m coords unchanged
      if (dim >= 3) {
        result[offset + 2] =
            iter.moveNext() ? iter.current : throw invalidCoordinates;
      }
      if (dim >= 4) {
        result[offset + 3] =
            iter.moveNext() ? iter.current : throw invalidCoordinates;
      }
      offset += dim;
    }

    return result;
  }
}
```

[WebMercatorConverter](https://github.com/navibyte/geospatial/blob/v0.12.0/dart/geobase/lib/src/utils/web_mercator_converter.dart)

```dart
import 'dart:math' as math;

import 'package:meta/meta.dart';

import '/src/constants/geodetic.dart';
import '/src/coordinates/geographic/geographic_functions.dart';
import '/src/tiling/convert/scaled_converter.dart';

// const _originShift = earthCircumferenceWgs84 / 2.0; // meters (~ 20037508.34)

/// Returns a clipped latitude in the range `[-89.999999, 89.999999]`.
///
/// As a special case if [lat] is `double.nan` then `double.nan` is returned.
double _clipLatitudeWithEpsilon(double lat) =>
    lat < -89.999999 ? -89.999999 : (lat > 89.999999 ? 89.999999 : lat);

/// A helper class to convert geographic coordinates to Web Mercator projection.
@internal
class WebMercatorConverter implements ScaledConverter {
  /// Create a converter from geographic coordinates to mercator projection.
  ///
  /// This implementation is based on the WGS 84 / Web Mercator projection
  /// ("EPSG:3857") aka "Pseudo-Mercator" or "Spherical Mercator".
  const WebMercatorConverter.epsg3857();

  /// The earth radius in meters.
  double get earthRadius => earthRadiusWgs84;

  /// The earth circumference in meters.
  double get earthCircumference => earthCircumferenceWgs84;

  /// The pixel ground resolution in meters at given [latitude] and map [size].
  double pixelResolutionAt(double latitude, num size) {
    final lat = latitude.clipLatitudeWebMercator();
    return (math.cos(lat * math.pi / 180.0) * earthCircumference) / size;
  }

  /// The map size from pixel ground [resolution] in meters at given [latitude].
  double sizeFromPixelResolutionAt(double latitude, double resolution) {
    final lat = latitude.clipLatitudeWebMercator();
    return (math.cos(lat * math.pi / 180.0) * earthCircumference) / resolution;
  }

  /// Converts geographic [longitude] to projected map x coordinate (metric).
  ///
  /// X origin at the prime meridian (lon: 0), X axis from west to east.
  double toProjectedX(double longitude) {
    final lon = longitude.wrapLongitude();
    return lon * earthCircumference / 360.0;
  }

  /// Converts geographic [latitude] to projected map y coordinate (metric).
  ///
  /// Y origin at the equator (lat: 0), Y from south to north.
  double toProjectedY(double latitude) {
    // clip latitude to range `[-89.999999, 89.999999]` to ensure no infinity
    final lat = _clipLatitudeWithEpsilon(latitude);

    final sinLat = math.sin(lat * math.pi / 180.0);
    final y = math.log((1.0 + sinLat) / (1.0 - sinLat)) / (4.0 * math.pi);
    return y * earthCircumference;
  }

  /// Converts projected map [x] coordinate (metric) to geographic longitude.
  ///
  /// X origin at the prime meridian (lon: 0), X axis from west to east.
  double fromProjectedX(double x) {
    return ((x / earthCircumference) * 360.0).wrapLongitude();
  }

  /// Converts projected map [y] coordinate (metric) to geographic latitude.
  ///
  /// Y origin at the equator (lat: 0), Y from south to north.
  double fromProjectedY(double y) {
    final y0 = y / earthCircumference;
    return (90.0 - 360.0 * math.atan(math.exp(-y0 * 2.0 * math.pi)) / math.pi)
        .clipLatitude();
  }

  /// Converts geographic [longitude] to x coordinate with range (0, [width]).
  ///
  /// X origin at the anti-meridian (lon: -180), X axis from west to east.
  @override
  double toScaledX(num longitude, {num width = 256}) {
    final lon = longitude.toDouble().wrapLongitude();
    return (0.5 + lon / 360.0) * width;
  }

  /// Converts geographic [latitude] to y coordinate with range (0, [height]).
  ///
  /// Y origin near the north pole (lat: 85.05112878), Y from north to south.
  @override
  double toScaledY(num latitude, {num height = 256}) {
    final lat = latitude.toDouble().clipLatitudeWebMercator();
    final sinLat = math.sin(lat * math.pi / 180.0);
    final y = 0.5 - math.log((1.0 + sinLat) / (1.0 - sinLat)) / (4.0 * math.pi);
    return y * height;
  }

  /// Converts [x] coordinate with range (0, [width]) to geographic longitude.
  ///
  /// X origin at the anti-meridian (lon: -180), X axis from west to east.
  @override
  double fromScaledX(num x, {num width = 256}) {
    final xc = x.clamp(0, width);
    return ((xc / width) - 0.5) * 360.0;
  }

  /// Converts [y] coordinate with range (0, [height]) to geographic latitude.
  ///
  /// Y origin near the north pole (lat: 85.05112878), Y from north to south.
  @override
  double fromScaledY(num y, {num height = 256}) {
    final yc = y.clamp(0, height);
    final y0 = 0.5 - yc / height;
    return 90.0 - 360.0 * math.atan(math.exp(-y0 * 2.0 * math.pi)) / math.pi;
  }
}
```
